### PR TITLE
Y2K Finance strategy

### PR DESCRIPTION
**Pull request by johnnyonline**

### CODE REVIEW

**Comment by johnnyonline** [comment]
Still need to add tests.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/BaseStrategy.sol [hunk 1]
```diff
@@ -84,6 +80,7 @@ abstract contract BaseStrategy is ReentrancyGuard, IStrategy {
     /// @inheritdoc IStrategy
     function isActive() public view virtual returns (bool) {
         if (isStrategiesActiveOverride) return false;
+        if (IERC20(assetVaultPrimaryAsset).balanceOf(address(this)) > 0) return true;
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/BaseStrategy.sol:83]: What is the definition of "active" with managed vaults? Epoch based strategies may not do anything even though the only purpose for inactivity is to deposit assets. Mapping this to Y2K, the weekly vaults are active between Wednesdays and Sunday, while Monday and Tuesday nothing happens other than people depositing into vaults and waiting for the epoch to start and vaults to trade actively.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol [hunk 2]
```diff
@@ -1,7 +1,126 @@
...
+        
+        (uint256 _id, uint256 _amount, uint256 _before, address _vault) = _getConfig(_configData);
+
+        _approve(assetVaultPrimaryAsset, _vault, _amount);
+        IY2KVault(_vault).deposit(_id, _amount, address(this));
+
+        uint256 _shares = IY2KVault(_vault).balanceOf(address(this), _id) - _before;
+
+        // TODO
+        // if (_getStakingContract(_vault, _id) != address(0)) _stake(_shares);
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol:81]: Staking still needs to be addressed.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol [hunk 3]
```diff
@@ -1,7 +1,126 @@
...
+        return _shares;
+    }
+
+    /// @dev Terminates the strategy - withdraw from fortGLP
+    /// @dev _configData:
+    /// @dev _index - index of the vault in the vaultFactory, determines the asset and the strike vault
+    /// @dev _amount - amount of the assetVaultPrimaryAsset to deposit, will be set to the balance if set to type(uint256).max
+    /// @dev _id - id of the vault in the vaultFactory, determines end date of the Epoch
+    /// @dev _type - true for Risk Vault, false for Hedge Vault
+    function terminate(bytes memory _configData) external override onlyManager returns (uint256) {
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol:92]: Do I understand it correctly that `execute` would be called before the Y2K epoch starts and `terminate` would be called after the Y2K epoch ended?

File: src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol [hunk 4]
```diff
@@ -1,7 +1,126 @@
...
+
+        IY2KVault(_vault).withdraw(_id, _amount, address(this), address(this));
+
+        return IY2KVault(_vault).balanceOf(address(this), _id) - _before;
+    }
+
+    /********************************** Internal Functions **********************************/
+
+    function _getConfig(bytes memory _configData) internal returns (uint256, uint256, uint256, address) {
+        (uint256 _index, uint256 _amount, uint256 _id, bool _type) = abi.decode(_configData, (uint256, uint256, uint256, bool));
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol:107]: Can we document how the RPC call expects the config data bytes? Is it basically just some JSON object with the respective field names and values?

File: src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVault.sol [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    /// @dev Returns the amount of tokens of token type `id` owned by `account`
+    /// @param account - address of the account to check the balance of
+    /// @param id - uint256 of the token type to check the balance of
+    function balanceOf(address account, uint256 id) external view returns (uint256);
+
+    /// @dev Mints Vault shares to receiver by depositing exact amount of assets
+    /// @param _assets - uint256 representing how many assets the user wants to deposit, a fee will be taken from this value
+    /// @param _receiver - address of the receiver of the assets provided by this function, that represent the ownership of the deposited asset
+    /// @param _id - uint256 in UNIX timestamp, representing the end date of the epoch. Example: Epoch ends in 30th June 2022 at 00h 00min 00sec: 1654038000
+    function deposit(uint256 _id, uint256 _assets, address _receiver) external;
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVault.sol:16]: Who is supposed to call `deposit`? Is it the vault manager or the Fortress user? I would have thought the user deposits into the managed vault, but then the user should not be concerned about the Y2K vault ID.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVaultFactory.sol [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+interface IY2KVaultFactory {
+
+    /// @dev Function the retrieve the addresses of the hedge and risk vaults, in an array, in the respective order
+    /// @param _index - uint256 of the market index which to the vaults are associated to
+    /// @return _vaults - Address array of two vaults addresses, [0] being the hedge vault, [1] being the risk vault
+    function getVaults(uint256 _index) external returns (address[] memory _vaults);
```
Comment (xh3b4sd) [inline src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVaultFactory.sol:9]: Not sure if there are any weird side effects of this but if we know there are only 2 items can we maybe safe gas here? 

```suggestion
    function getVaults(uint256 _index) external returns (address[2] memory _vaults);
```

File: src/arbitrum/epoch-based-managed-vaults/stratagies/BaseStrategy.sol (hunk 1 above)
Comment (johnnyonline) [reply src/arbitrum/epoch-based-managed-vaults/stratagies/BaseStrategy.sol:83]: Definition of "active" is if value is in the vault. If you look at it from the opposite direction, an epoch can be finished only when no AssetVault (and so also StrategyVault) is active, i.e. when all assets/value is back in the MetaVault.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol (hunk 3 above)
Comment (johnnyonline) [reply src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol:92]: Yeah I think so, once I implement tests we'll see it in practice.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol (hunk 4 above)
Comment (johnnyonline) [reply src/arbitrum/epoch-based-managed-vaults/stratagies/Y2KFinanceStrategy.sol:107]: Oh yeah. We'll need good docs for this one. It's basically a bunch of data that you encode and pass to the contract, no biggie.

File: src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVault.sol (hunk 5 above)
Comment (johnnyonline) [reply src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVault.sol:16]: The manager of the AssetVault (which initially is the manager of the MetaVault, can be updated).

File: src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVaultFactory.sol (hunk 6 above)
Comment (johnnyonline) [reply src/arbitrum/epoch-based-managed-vaults/stratagies/interfaces/IY2KVaultFactory.sol:9]: The function returns a dynamic array. See (https://arbiscan.io/address/0x984e0eb8fb687afa53fc8b33e12e04967560e092#code#F1#L602).

**Review by xh3b4sd — APPROVED** [review]
I made some notes for our next walkthrough. This PR is fine for now from my PoV. Thanks for answering my questions.

**Comment by johnnyonline** [comment]
Before merging, need to execute tests in accordance with Y2K timelines, then disable them.

**Pull request merged**